torchreid/engine/image/am_softmax.py

```diff
@@ -73,181 +68,93 @@ def __init__(self, datamanager, models, optimizers, reg_cfg, metric_cfg, schedul
-                model_loss, model_loss_summary, model_avg_acc, model_logits = self._single_model_losses(
-                    self.models[model_name], train_records, imgs, obj_ids, n_iter, model_name, num_packages
-                )
-
-                avg_acc += model_avg_acc / float(num_models)
-                total_loss += model_loss / float(num_models)
+            for i, model_name in enumerate(model_names):
+                loss, model_loss_summary, acc, scaled_logits = self._single_model_losses(
+                    self.models[model_name], imgs, targets, n_iter, model_name
+                    )
+                models_logits[i] = scaled_logits
                 loss_summary.update(model_loss_summary)
+                if i == 0: # main model
+                    main_acc = acc
 
-                for trg_id in range(self.num_targets):
-                    if model_logits[trg_id] is not None:
-                        out_logits[trg_id].append(model_logits[trg_id])
-
-            if len(model_names) > 1:
-                num_mutual_losses = 0
-                mutual_loss = torch.zeros([], dtype=imgs.dtype, device=imgs.device)
-                for trg_id in range(self.num_targets):
-                    if len(out_logits[trg_id]) <= 1:
-                        continue
-
-                    with torch.no_grad():
-                        trg_probs = torch.softmax(torch.stack(out_logits[trg_id]), dim=2).mean(dim=0)
-
-                    for model_id, logits in enumerate(out_logits[trg_id]):
-                        log_probs = torch.log_softmax(logits, dim=1)
-                        m_loss = (trg_probs * log_probs).sum(dim=1).mean().neg()
+            for i, model_name in enumerate(model_names):
+                self.optims[model_name].zero_grad()
+                if num_models > 1: # mutual learning
+                    mutual_loss = 0
+                    for j in range(num_models):
+                        if i != j:
+                            with torch.no_grad():
+                                trg_probs = F.softmax(models_logits[j], dim=1)
+                            mutual_loss += self.loss_kl(F.log_softmax(models_logits[i], dim = 1),
+                                                    trg_probs)
+                    loss_summary[f'mutual_{model_names[i]}'] = mutual_loss.item()
 
-                        mutual_loss += m_loss
-                        loss_summary['mutual_{}/{}'.format(trg_id, model_names[model_id])] = m_loss.item()
-                        num_mutual_losses += 1
+                    should_turn_off_mutual_learning = self._should_turn_off_mutual_learning(self.epoch)
+                    coeff_mutual_learning = int(not should_turn_off_mutual_learning)
 
-                should_turn_off_mutual_learning = self._should_turn_off_mutual_learning(self.epoch)
-                coeff_mutual_learning = int(not should_turn_off_mutual_learning)
+                    loss += coeff_mutual_learning * mutual_loss / (num_models - 1)
```

Comment:
Why do we need compute the mutual loss if `should_turn_off_mutual_learning` is `False`?

Reply:
Now, it is seems to be okay